content/rs/developing/crdbs/developing-streams-crdb.md

```diff
@@ -1,278 +1,320 @@
-| t1   | XADD x 100-1 f1 v1              | XADD x 100-2 f1 v1              |
-| t2   | XRANGE x - + --> [100-1]        | XRANGE x - + --> [100-2]        |
-| t3   | --- sync ---                    | --- sync ---                    |
-| t4   | XRANGE x - + --> [100-1, 100-2] | XRANGE x - + --> [100-1, 100-2] |
+Notice also that the synchronized streams contain no duplicate IDs. As long as you allow the database to generate your stream IDs, you'll never have more than one stream entry with the same ID.
 
-### Implementation notes
+{{% note %}}
+Open source Redis uses one radix tree (referred to as `rax` in the code base) to implement each stream. However, Active-Active databases implement a single logical stream using one `rax` per region.
+Each region adds entries only to its associated `rax` (but can remove entries from all `rax` trees).
+This means that XREAD and XREADGROUP iterate simultaneously over all `rax` trees and return the appropriate entry by comparing the entry IDs from each `rax`.
+{{% /note %}}
 
-OSS Redis uses one radix tree (rax) to hold the entries, but an Active-Active stream holds a radix tree per-writing location.
-Every location can add entries only to its rax but can remove entries from all rax trees.
-XREAD iterates simultaneously on all rax trees and returns the appropriate entry in each loop by comparing entry-IDs from the different rax trees.
+### Conflict resolution
 
-### The observed-remove approach
+Active-Active databases use an "observed-remove" approach to automatically resolve potential conflicts.
 
-In Active-Active databases, the observed-remove approach is a way to auto-resolve conflicts.
-For stream data operations, we use the observed-remove approach.
-It means that in case of a concurrent operation to DEL, DEL only affects data that currently exists.
+With this approach, a delete only affects the locally observable data.
 
-For example:
+In the example below, a stream, `x`, is created at _t1_. At _t3_, the stream exists in two regions.
+
+<table style='width: 100%;'>
+<thead>
+<tr>
+<th>Time</th>
+<th>Region 1</th>
+<th>Region 2</th>
+</tr>
+</thead>
+
+<tbody>
+<tr>
+<td><em>t1</em></td>
+<td><code>XADD messages * text hello</code></td>
+<td></td>
+</tr>
+
+<tr>
+<td><em>t2</em></td>
+<td><em>— Sync —</em></td>
+<td><em>— Sync —</em></td>
+</tr>
 
-| Time | Location 1               | Location 2               |
-| ---- | ------------------------ | ------------------------ |
-| t1   | XADD x 100-1 f1 v1       |                          |
-| t2   | --- sync ---             | --- sync ---             |
-| t3   | XRANGE x - + --> [100-1] | XRANGE x - + --> [100-1] |
-| t4   | DEL x                    | XADD x 100-2 f1 v1       |
-| t5   | --- sync ---             | --- sync ---             |
-| t6   | XRANGE x - + --> [100-2] | XRANGE x - + --> [100-2] |
+<tr>
+<td><em>t3</em></td>
+<td><code>XRANGE messages - +</code> <br/><strong>→ [1589929244828-2100]</strong></code></td>
```

Comment:
One more thing - 
The SEQ part (which is 2100, 3700 here) should be the `SEQ == instance_number%32`. So I would expect 1,2.
Moveover - their %32 for both is 20 which can't be in strict mode (this would mean they are both the same instance)

Reply:
@AlonMagrafta Not sure what you mean by: "their %32 for both is 20" where do you see this?

Reply:
What I meant was
2100%32 = 20 as well as 3700%32 = 20
Which indicates both are the same region. But I see you updated it to be 1 and 2 which is the right